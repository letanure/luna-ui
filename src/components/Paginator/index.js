import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Icon from '../Icon';
import Hint from '../Hint';

/**
 * Paginator component.
 * E.g.:
 *
 * ```javascript
 * paginationInfo : {
 *  size: 3,
 *  last: false,
 *  first: true,
 *  number: 0,
 *  totalPages: 3,
 *  totalElements: 25,
 *  numberOfElements: 25
 * }
 * ```
 *
 * ```html
 * <Paginator onGoToPage={this.goToPage.bind(this)} paginationInfo={paginationInfo} />
 * ```
*/
class PaginatorComponent extends React.PureComponent {
  goToPreviousPage() {
    if (this.props.paginationInfo.first) {
      return;
    }

    this.props.onGoToPage(this.props.paginationInfo.number - 1);
  }

  goToNextPage() {
    if (this.props.paginationInfo.last) {
      return;
    }

    this.props.onGoToPage(this.props.paginationInfo.number + 1);
  }

  getPaginator() {
    const { className } = this.props;

    const ControlButton = styled.button`
      color: #fff;
      border: none;
      outline: none;
      cursor: pointer;
      font-weight: bold;
      border-radius: 2px;
      text-transform: uppercase;
      transition: opacity linear 0.25s;

      margin: 0 5px;

      height: 23px;
      padding: 0 8px;

      background-color: ${(props) => props.theme.buttonColors.primary};

      &:hover {
        opacity: 0.8;
      }

      &:disabled {
        opacity: 0.4;
        cursor: not-allowed;
      }
    `;

    return (
      <div className={className}>
        <ControlButton
          disabled={this.props.paginationInfo.first}
          onClick={() => this.goToPreviousPage()}>
          <Icon name="angle-left" />
        </ControlButton>
        <Hint>{this.props.paginationInfo.number + 1} - {this.props.paginationInfo.totalPages}</Hint>
        <ControlButton
          disabled={this.props.paginationInfo.last}
          onClick={() => this.goToNextPage()}>
          <Icon name="angle-right" />
        </ControlButton>
      </div>
    );
  }

  render() {
    const showPaginator = this.props.paginationInfo.totalPages > 1;

    if (showPaginator) {
      return this.getPaginator();
    }

    return null;
  }
}

PaginatorComponent.propTypes = {
  paginationInfo: PropTypes.shape({
    size: PropTypes.number,
    last: PropTypes.bool,
    first: PropTypes.bool,
    number: PropTypes.number,
    totalPages: PropTypes.number,
    totalElements: PropTypes.number,
    numberOfElements: PropTypes.number
  }).isRequired,
  onGoToPage: PropTypes.func.isRequired,
  className: PropTypes.string
};

PaginatorComponent.defaultProps = {
  paginationInfo: {
    size: 0,
    last: false,
    first: false,
    number: 0,
    totalPages: 0,
    totalElements: 0,
    numberOfElements: 0
  },
  onGoToPage: () => true,
  className: ''
};

export default PaginatorComponent;
