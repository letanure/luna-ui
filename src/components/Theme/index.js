import React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';

import defaultTheme from '../../config/defaultTheme';
import DefaultStyles from '../../config/defaultStyles';
import DefaulTyphography from '../../config/defaultTyphography';

/**
 * Theme component.
 * E.g.:
 * ```html
 * <Theme theme={{...}}>
 *  ...
 * </Theme>
 * ```
 */
class ThemeComponent extends React.PureComponent {
  render() {
    const { theme, children } = this.props;

    return (
      <ThemeProvider theme={theme}>
        <div>
          <DefaultStyles />
          <DefaulTyphography />
          {children}
        </div>
      </ThemeProvider>
    );
  }
}

ThemeComponent.propTypes = {
  theme: PropTypes.object,
  children: PropTypes.node
};

ThemeComponent.defaultProps = {
  theme: defaultTheme,
  children: null
};

export default ThemeComponent;
