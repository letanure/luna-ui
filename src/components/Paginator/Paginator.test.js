import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import Theme from '../../config/defaultTheme';
import ThemeProvider from '../Theme/index';

import Paginator from './index';

describe('<Paginator />', () => {
  let component; let onGoToPage; let paginationInfo;
  const generatePaginationInfo = (size, last, first, number, totalPages, totalElements, numberOfElements) => {
    return { size, last, first, number, totalPages, totalElements, numberOfElements };
  };

  describe('NextPage and Previous page disabled', () => {
    beforeEach(() => {
      onGoToPage = jest.fn();
      paginationInfo = generatePaginationInfo(3, false, true, 0, 3, 25, 25);

      component =
        <ThemeProvider theme={Theme}>
          <Paginator onGoToPage={onGoToPage} paginationInfo={paginationInfo} />
        </ThemeProvider>;
    });

    test('HandleClick: NextPage', () => {
      const wrapper = mount(component);
      wrapper.find('button').at(1)
        .simulate('click');

      expect(onGoToPage).toBeCalledWith(1);
    });

    test('HandleClick: Previous page disabled', () => {
      const wrapper = mount(component);

      wrapper.find('button').at(0)
        .simulate('click');
      expect(onGoToPage).not.toBeCalled();
    });

    test('toMatchSnapshot', () => {
      const tree = renderer
        .create(component)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('PreviousPage and Next page disabled', () => {
    beforeEach(() => {
      onGoToPage = jest.fn();
      paginationInfo = generatePaginationInfo(3, true, false, 3, 3, 25, 25);

      component =
        <ThemeProvider theme={Theme}>
          <Paginator onGoToPage={onGoToPage} paginationInfo={paginationInfo} />
        </ThemeProvider>;
    });

    test('HandleClick: PreviousPage', () => {
      const wrapper = mount(component);
      wrapper.find('button').at(0)
        .simulate('click');

      expect(onGoToPage).toBeCalledWith(2);
    });

    test('HandleClick: NextPage page disabled', () => {
      const wrapper = mount(component);

      wrapper.find('button').at(1)
        .simulate('click');
      expect(onGoToPage).not.toBeCalled();
    });

    test('toMatchSnapshot', () => {
      const tree = renderer
        .create(component)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('TotalPages equal 0', () => {
    beforeEach(() => {
      onGoToPage = jest.fn();
      paginationInfo = generatePaginationInfo(0, true, true, 0, 0, 0, 0);

      component =
        <ThemeProvider theme={Theme}>
          <Paginator onGoToPage={onGoToPage} paginationInfo={paginationInfo} />
        </ThemeProvider>;
    });

    test('Component should not load', () => {
      const wrapper = mount(component);

      expect(wrapper.html()).toBe('<div></div>');
    });
  });
});
