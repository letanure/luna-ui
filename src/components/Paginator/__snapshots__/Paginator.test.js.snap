// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Paginator /> NextPage and Previous page disabled toMatchSnapshot 1`] = `
<div>
  <div
    className=""
  >
    <button
      className="sc-ifAKCX eCgCfk"
      disabled={true}
      onClick={[Function]}
    />
    <span
      className="sc-EHOje kJmRE"
    >
      1
       - 
      3
    </span>
    <button
      className="sc-ifAKCX eCgCfk"
      disabled={false}
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`<Paginator /> PreviousPage and Next page disabled toMatchSnapshot 1`] = `
<div>
  <div
    className=""
  >
    <button
      className="sc-iwsKbI bvTUrg"
      disabled={false}
      onClick={[Function]}
    />
    <span
      className="sc-gZMcBi fuUimc"
    >
      4
       - 
      3
    </span>
    <button
      className="sc-iwsKbI bvTUrg"
      disabled={true}
      onClick={[Function]}
    />
  </div>
</div>
`;
